### PR TITLE
folder move issues

### LINKED ISSUE #79: API fails to move folders properly on S3
## Description

When attempting to move a folder into some sub folder, the user is met with a success message, though the files don't actually get moved properly. This behavious seems to only happen with an S3 configured bucket

### PR DESCRIPTION
**Pull request by joaxaca**
## Summary
closes #79
Fixes issues with moving folders with content, nested folders and files to different locations in the media manager.

## Testing Checklist
Test moving folders around. With or without contents.
    - move an empty folder
    - move a folder with a file
    - move a folder with an empty folder
    - move a folder with a folder with a file
    - move a folder that has the same name as a folder in another level

### CODE REVIEW

**Review by m-triassi — CHANGES_REQUESTED** [review]
did a first pass review, mostly spotted some linting issues, but and some missing tests.

One other thing i noticed is the title is a default title, and the description references an internal repo / issue, lets remove those references and describe this change more accurately. This PR concerns issues with moving folders on S3 file systems

Generally though it don't see any issues with the logic and code here, so if the small issues outlined are resolved we'll probably be good to go!

File: src/Http/Controllers/MediaManagerController.php [hunk 1]
```diff
@@ -75,15 +72,32 @@ public function update(Request $request)
...
+
+        if ( $files->count() > 0 ) {
+            $files->each(
+                function ($media) use ($source, $destination, $moved)
+                {
+                    $directory = trim(str_replace($source, $destination, $media->directory), '/');
+                    $media->move($directory);
+                    $moved[] = $media->fresh();
+                }
+            );
```
[the remark this answers is not in the document]
Comment (a-drew) [reply src/Http/Controllers/MediaManagerController.php:92]: auto lint anyone?

**Review by m-triassi — CHANGES_REQUESTED** [review]
Hm, After testing this locally, it doesn't seem to be preserving the folder structure or inner files? Maybe i'm doing something wrong here: 

https://user-images.githubusercontent.com/9440691/181105855-4af4b3e9-91a1-4a2f-8bcd-c1ee9a102e2d.mov

Would you be able to show me an example video of it working as intended? maybe its my local set up

**Comment by joaxaca** [comment]
https://user-images.githubusercontent.com/19713336/181287065-bc0095e0-e6c4-4134-bff5-332170566ed5.mov

Maybe a `composer update` is needed on the TCE project?

**Review by m-triassi — APPROVED** [review]
Thanks for your patience, i've been sort of dragging my feet on this one when i couldn't get it to work locally properly. I'm still running into that issue but it seems to work fine on your local. From a code stand point i can't see anything i disagree with, so maybe @a-drew you can verify functionality on this locally?

**Review by a-drew — CHANGES_REQUESTED** [review]
hey guys, here's a list of scenarios I've tested with and the results:

- move an empty folder ❌
   - empty folder gets deleted eg: /a/b => /a/b or to /a/c/b
- move a folder with a file ✅
   - 1 level to 1 level ✅
   - 1 level to deeper level ✅
   - deeper level to 1st level  ✅
   - test with multiple files ✅
   - move to same location fails ❌
         - error: file already exists.
         - no frontend indication of fail
- move a folder with an empty folder ❌
  - the empty subfolder is deleted
  - this happens to nested empty folders too, only parent survives the move.
     so if I do /a/1/2/3 => /b/1/2/3, I only get /b/1
- move a folder with a folder with a file 
   - with 1 file ✅
   - with nested files and folders ✅
- move a folder that has the same name as a folder in another level 
  - folders without conflicting files get merged succesfully ✅
  - if files are conflicting, the move crashes / stops at the offending file ❌

Scenarios not covered in testing checklist and other notes:
- move folder with a folder with a file ✅
- cant move anything to root level.
  - not very important tbh
- move file to parent folder ✅
- move file to nested folder ❌
   - translation error but media entry was changed
   - cache clear updates location
- clicking move when inside a target folder instead of selecting it in the ui will try to move to existing path eg: /a/b => /a/b

**Review by a-drew — COMMENTED** [review]
Thanks for handling the leftover issues. The empty folders are now preserved during moves.
For the file collisions, I liked what you did with the exceptions. Now should we attempt to merge folders and keep the latest files during collisions using their timestamps? Or do we get some frontend error handling for those new exceptions and close this PR?

cc: @m-triassi

**Review by a-drew — APPROVED** [review]

**Comment by m-triassi** [comment]
Honestly, i think it'd be fine to just give them an error and say the folder already exists in the destination

We _could_ do the thing where we rename the folder to preserve the original file structure and not have to recurse down the whole file tree of a particular folder to handle merges, sort of how we handle files with the same names, so basically we can have this:

```
folder_a
   sub_a
   sub _b
sub_a
```

when you move `sub_a` into `folder_a`

it would become 
```
folder_a
   sub_a
   sub _a_1
   sub_b
```

Basically count how many folders within the destination start with the sub string `sub_a` (or what ever the target name is) and then append that count to the end of the folder name. Not the _best_ solution, but honestly i'd say the least painful solution that covers the edge case atleast

**Comment by joaxaca** [comment]
I also vote for the option of returning an error. 

There could be a question asking if the user wants to proceed and merge. That way the user is on control at all times.

Note that there are a couple of bugs that were fixed when I wrote the tests on the `374-add-file-manipulation-tests` branch.

**Comment by a-drew** [comment]
alright, let's go with the frontend error option for now and if it becomes relevant again we can revisit renaming / merging.
